tephra-hbase-compat-0.98/src/main/java/co/cask/tephra/hbase98/coprocessor/TransactionProcessor.java

```diff
@@ -151,6 +156,35 @@ public void preGetOp(ObserverContext<RegionCoprocessorEnvironment> e, Get get, L
   }
 
   @Override
+  public void preDelete(ObserverContext<RegionCoprocessorEnvironment> e, Delete delete, WALEdit edit,
+                        Durability durability) throws IOException {
+    // Translate deletes into our own delete tombstones
+    // Since HBase deletes cannot be undone, we need to translate deletes into special puts, which allows
+    // us to rollback the changes (by a real delete) if the transaction fails
+
+    // Deletes that are part of a transaction rollback do not need to be handled
```

Comment:
you may add

"do not need to be handled specially: they do actual delete of data and will never be rolled back"